### ISSUE TITLE
Rewards/Penalties (Slashing Conditions) in the Beacon Chain

### ISSUE THREAD

**Issue by nisdas** [prysmaticlabs/prysm#259]
In order to implement this we will need to 
- Compute the total deposits of every validator who participated in the last epoch 
- Compute the online reward and offline penalty based on Casper FFG Incentive(not specified yet)
- Add the online reward to every validator that participated in the last epoch and subtract the penalty from those who did not

Rewards for crosslinks are not in scope for Ruby so will be pushed to the next release.
Also slashing conditions , ex: penalties for signing or attesting to non-canonical-chain blocks ,etc have not been fully specified yet. so we can stub them out until it gets elaborated on in the research spec

**Comment by nisdas** [comment]
I think I will hold off on this until slashing conditions are specified and the spec is slightly more mature

**Comment by terenc3t** [comment]
Here is the latest spec:
```
Calculate rewards for FFG votes
If total_attester_deposits * 3 >= total_deposits * 2, set crystallized_state.justified_epoch to equal crystallized_state.current_epoch (note: this is still the previous epoch at this point in the computation). If this happens, and the justified epoch was previously crystallized_state.current_epoch - 1, set the crystallized_state.finalized_epoch to equal that value.
Compute the online_reward and offline_penalty based on the Casper FFG incentive and quadratic leak rules (not yet fully specified)
Add the online_reward to every validator who participated in the last epoch, and subtract the offline_penalty from everyone who did not
Set attester_bitfield to be all zeroes, with length ceil_div8(len(active_validators)). Set total_attester_deposits to 0
```

**Comment by nisdas** [comment]
Hey @terenc3t , isnt the v2.1 spec not finalised yet ? I think in the sharding gitter channel they said it was most likely suspect to heavy change.

**Comment by nisdas** [comment]
Closed via #323
